Sources/ScheduledTasks.php

```diff
@@ -1542,4 +1542,196 @@ function scheduled_remove_old_drafts()
+ * pruning log_topics.
+ */
+function scheduled_prune_log_topics()
+{
+	global $smcFunc, $sourcedir, $modSettings;
+
+	if (empty($modSettings['mark_read_max_users']))
+		return true;
+
+	// Convert to timestamps for comparison
+	$markReadCutoff = time() - $modSettings['mark_read_beyond'] * 86400;
+	$cleanupBeyond = time() - $modSettings['mark_read_delete_beyond'] * 86400;
+	$maxMembers = $modSettings['mark_read_max_users'];
+
+	// Try to prevent timeouts
+	@set_time_limit(300);
+	if (function_exists('apache_reset_timeout'))
+		@apache_reset_timeout();
+
+	// Start off by finding the records in log_boards, log_topics & log_mark_read
+	// for users who haven't been around the longest...
+
+	// Find the oldest $maxMembers in log_boards
+	$sql = "SELECT lb.id_member, m.last_login
+		FROM {db_prefix}members m
+		INNER JOIN
+		(
+			SELECT DISTINCT id_member
+			FROM {db_prefix}log_boards
+		) lb
+		ON lb.id_member = m.id_member
+		WHERE m.last_login <= {int:cutoff}
+		ORDER BY m.last_login
+		LIMIT {int:limit}";
+	$result = $smcFunc['db_query']('', $sql,
+		array(
+			'limit' => $maxMembers,
+			'cutoff' => $cleanupBeyond,
+		)
+	);
+	$members = array();
+	while($row = $smcFunc['db_fetch_assoc']($result))
+		$members[$row['id_member']] = $row['last_login'];
+	$smcFunc['db_free_result']($result);
+
+	// Find the oldest $maxMembers in log_mark_read
+	$sql = "SELECT lmr.id_member, m.last_login
+		FROM {db_prefix}members m
+		INNER JOIN
+		(
+			SELECT DISTINCT id_member
+			FROM {db_prefix}log_mark_read
+		) lmr
+		ON lmr.id_member = m.id_member
+		WHERE m.last_login <= {int:cutoff}
+		ORDER BY m.last_login
+		LIMIT {int:limit}";
+	$result = $smcFunc['db_query']('', $sql,
+		array(
+			'limit' => $maxMembers,
+			'cutoff' => $cleanupBeyond,
+		)
+	);
+	// Using id_member as key basically eats dupes...
+	while($row = $smcFunc['db_fetch_assoc']($result))
+		$members[$row['id_member']] = $row['last_login'];
+	$smcFunc['db_free_result']($result);
+
+	// Find the oldest $maxMembers in log_topics
+	// Note the different cutoff, as we may need to mark these as read
+	$sql = "SELECT lt.id_member, m.last_login
+		FROM {db_prefix}members m
+		INNER JOIN
+		(
+			SELECT DISTINCT id_member
+			FROM {db_prefix}log_topics
+			WHERE unwatched = {int:unwatched}
+		) lt
+		ON lt.id_member = m.id_member
+		WHERE m.last_login <= {int:cutoff}
+		ORDER BY m.last_login
+		LIMIT {int:limit}";
+	$result = $smcFunc['db_query']('', $sql,
+		array(
+			'limit' => $maxMembers,
+			'cutoff' => $markReadCutoff,
+			'unwatched' => 0,
+		)
+	);
+	// Using id_member as key basically eats dupes...
+	while($row = $smcFunc['db_fetch_assoc']($result))
+		$members[$row['id_member']] = $row['last_login'];
+	$smcFunc['db_free_result']($result);
+
+	// Sort by last_login & limit to $maxMembers entries
+	asort($members);
+	$members = array_slice($members, 0, $maxMembers, TRUE);
+
+	// Either mark boards as read or purge entries, based on last_login
+	foreach($members as $id => $last_login)
```

Comment:
To run for every member throw the delete 3 delete commands is not effience,
you should rather collect all member and use the `` IN ({array_int:members)`` style
for exp.
```
			// Delete rows from log_boards
			$sql = 'DELETE FROM {db_prefix}log_boards
				WHERE id_member in ( {array_int:members} )';
			$smcFunc['db_query']('', $sql,
				array(
					'members' => $ids,
				)
			);
```

Reply:
Will test, & if faster, will do.  

This could all be collapsed into just 4 queries per execution - 3 LARGE deletes & 1 LARGE insert...

Reply:
dunno how many entries this delete array can have,
alternative you could do some kind of batch size,
for exp. fire the the delete statements when the array size is 500 or the last run,
instead of to collect them all.

Reply:
I am testing restructures of this logic, and performance improvements vary depending on environment, etc.  But I've been seeing _4x - 10x_ improvements in speed.  

Good input.